### PR TITLE
feat: add `ndarray/shift`

### LINKED ISSUE #84: [RFC]: add `ndarray/shift`
ref: https://github.com/stdlib-js/stdlib/issues/2656

### PR DESCRIPTION
**Pull request by headlessNode**
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/84.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `ndarray/shift`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/84
-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/ndarray/shift/README.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+The function accepts the following arguments:
+
+-   **x**: input ndarray.
+-   **options**: function options (_optional_).
+
+The function supports the following `options`:
+
+-   **dim**: dimension along which to perform the operation. If provided an integer less than zero, the dimension index is resolved relative to the last dimension, with the last dimension corresponding to the value `-1`. Default: `0`.
+
+By default, the function performs operation along the first dimension of the input [`ndarray`][@stdlib/ndarray/ctor]. To perform operation along any other input [`ndarray`][@stdlib/ndarray/ctor], provide the `dim` option.
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/shift/README.md]:
```suggestion
By default, the function performs operation along the first dimension of the input [`ndarray`][@stdlib/ndarray/ctor]. To perform operation along a desired dimension, provide the `dim` option.
```

File: lib/node_modules/@stdlib/ndarray/shift/lib/main.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/**
+* Returns an array containing a read-only truncated view of an input ndarray and a read-only view of the first element(s) along a specific dimension.
+*
+* @param {ndarray} x - input array
+* @param {Object} [options] - function options
+* @param {integer} [options.dim=0] - dimension along which to perform the operation
+* @throws {TypeError} first argument must be an ndarray having one or more dimensions
+* @throws {TypeError} options argument must be an object
+* * @throws {TypeError} must provide valid options
+* @throws {RangeError} dimension index exceeds the number of dimensions
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/shift/lib/main.js]:
```suggestion
```

**Comment by kgryte** [comment]
@headlessNode There are likely various changes from https://github.com/stdlib-js/stdlib/pull/8145 that need to be carried over to this PR.

File: lib/node_modules/@stdlib/ndarray/shift/README.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# shift
+
+> Return an array containing a read-only truncated view of an input [`ndarray`][@stdlib/ndarray/ctor] and a read-only view of the first element(s) along a specific dimension.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/ndarray/shift/README.md]:
```suggestion
> Return an array containing a read-only truncated view of an input [`ndarray`][@stdlib/ndarray/ctor] and a read-only view of the first element(s) along a specified dimension.
```

Here and everywhere else.

Comment (kgryte) [reply lib/node_modules/@stdlib/ndarray/shift/README.md]: The reason why `specified` is still appropriate is that, even if a user does not explicitly provide a `dim` option, the option is still implicitly specified (i.e., omitting the `dim` option is the same as explicitly specifying `-1`).

File: lib/node_modules/@stdlib/ndarray/shift/test/test.js [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+
+	buf = zeroTo( 8, 'float64' );
+	sh = [ 2, 4 ];
+	st = [ 4, 1 ];
+	o = 0;
+	ord = 'row-major';
+
+	x = new ndarray( 'float64', buf, sh, st, o, ord );
+	opts = {
+		'dim': 1
```
Comment (kgryte) [inline lib/node_modules/@stdlib/ndarray/shift/test/test.js]: @headlessNode Always test to the interface, not the implementation. As I did in https://github.com/stdlib-js/stdlib/pull/8145, if we document that we support negative dimension indices, we need to include tests for negative dimension indices. When writing tests, it can be useful to ask, "If I didn't know anything about the implementation and only read the documentation and function signature, what behavior would I expect to observe?" When answering that question, you will almost always add more tests and be more comprehensive in your testing than you are now.

File: lib/node_modules/@stdlib/ndarray/shift/README.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+## Examples
+
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var array = require( '@stdlib/ndarray/array' );
+var ndarray2array = require( '@stdlib/ndarray/to-array' );
+var zeroTo = require( '@stdlib/array/zero-to' );
+var pop = require( '@stdlib/ndarray/pop' );
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/shift/README.md]:
```suggestion
var shift = require( '@stdlib/ndarray/shift' );
```

File: lib/node_modules/@stdlib/ndarray/shift/README.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+```
+
+The function accepts the following arguments:
+
+-   **x**: input ndarray.
+-   **options**: function options (_optional_).
+
+The function supports the following `options`:
+
+-   **dim**: dimension along which to perform the operation. If provided an integer less than zero, the dimension index is resolved relative to the last dimension, with the last dimension corresponding to the value `-1`. Default: `0`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/ndarray/shift/README.md]: The default should be `-1`, not `0`. The nature of the operation does not have any bearing on the `dim` option. Meaning, what value the `dim` option is is completely orthogonal to the fact that the function removes the first element(s). We should be consistent with other APIs and have the default dimension be `-1`. That affects the entirety of this PR.

**Review by kgryte — CHANGES_REQUESTED** [review]
Main comment is that this package needs to be refactored to use a different default for `dim`.

**Review by kgryte — APPROVED** [review]
LGTM after clean-up.
